### PR TITLE
Feature/overview widgets

### PR DESCRIPTION
**Pull request by adamjohnwright**
With this release there are three categories of widgets: ontology-browser, disease and overview. 

Ontology-browser is very bare bones widget and has been added to pages:
- phenotype
- life_stage
- go_term
- do_term
- anatomy_term

Disease
This widget is on the Gene page and is complete. Will need to be tested by curators.

Overview widgets. 
I created overview widgets for all 40 pages. This is not to say that all will be released. I have commented out more than half of them. The ones that are not commented out are (total 17): 
- analysis
- anatomy_term
- antibody
- expression_cluster
- feature
- gene (already made) 
- go_term
- laboratory
- microarray_results
- operon
- person
- rnai
- strain
- transposon
- transposon_family
- variation
- wbprocess

Ones that are close and just require some little bit more work are (total 13):
- construct
- homology_group
- interaction
- molecule
- paper
- pcr-product
- phenotype
- position_matrix
- pseudogene
- rearrangement
- sequence
- transcript 
- transgene

Difference results from AceDB:
* This is where I think I figured out the logic but the answers do not match.
- do_term
- gene_class

Missing homology fields
- motif
- protein
- structure_data

Missing fields other (might be because of homology or other)
- clone
- gene_cluster
- life_stage

I have removed unused namespaces with the command: lein eastwood '{:add-linters [:unused-namespaces]}'

I split the generic.clj file into to files generic_functions.clj and generic_fields.clj

generic_fields.clj contains fields that are found in more than one widget and could be made generic

generic_functions.clj contains only one function right now and probably should be moved somewhere. I just haven't had time to figure out where yet. 

I am planning on having this pull request pulled in and then I will look over to see if I can see any differences between prod and staging. If I see any differences I will either fix the issue or comment out the widget. Once that stage is complete I will let the curators know about the new widgets and see if they could approve them. After that I would like to push a round of widgets into production. After that I will try to finish off all the widgets that I have listed as close. 

I know this is a lot of code to look over. That is why I have decided to stop here and give us a chance to pull it is in before finishing more of the widgets off. Thanks for every ones help in advance. :)

### CODE REVIEW

File: src/rest_api/classes/do_term/widgets/overview.clj [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                      (remove
+                        nil?
+                        (for [dh dhs
+                              :let [source-db (:database/id (:gene.database/database dh))
+                                    source-type  (:database-field/id (:gene.database/field dh))]]
+                          (if (and
+                                (= source-db "OMIM")
+                                (= source-type "gene"))
+                            (:gene.database/accession dh)))))]
+            (if (some? id)
```
Comment (sibyl229) [inline src/rest_api/classes/do_term/widgets/overview.clj]: maybe (if id ....) ?

File: src/rest_api/classes/anatomy_term/widgets/overview.clj [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+           [?af :anatomy-function/not-involved ?nih]])
+
+(def q-anatomy-term-to-expression-cluster
+  '[:find ?e
+    :in $ ?a
+    :where [?eh :expression-cluster.anatomy-term/anatomy-term ?a]
+           [?e :expression-cluster/anatomy-term ?eh]])
+
+(defn transgenes [anatomy-term]
+  {:data (if-let [expression-patterns (:anatomy-term/expr-descendent anatomy-term)]
```
Comment (mgrbyte) [inline src/rest_api/classes/anatomy_term/widgets/overview.clj]: when-let maybe?

File: src/rest_api/classes/anatomy_term/widgets/overview.clj [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                                db (:db/id anatomy-term))]
+             (for [eid eids
+                   :let [expression-cluster (d/entity db (first eid))]]
+               {:expession_cluster (pack-obj expression-cluster)
+                :description (first (:expression-cluster/description expression-cluster))}))
+     :description "expression cluster data"}))
+
+(defn term [anatomy-term]
+  {:data (if-let [holder (:anatomy-term/term anatomy-term)]
+             {:id (:anatomy-term.term/text holder)
```
Comment (mgrbyte) [inline src/rest_api/classes/anatomy_term/widgets/overview.clj]: maybe pull out dup call `(:anatomy-term.term/text holder)` into a var in a let expression?

File: src/rest_api/classes/antibody/widgets/overview.clj [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   :description "the corresponding gene the antibody was generated against"})
+
+(defn antigen [antibody]
+  {:data {:comment (if-let [antigen (:antibody/antigen antibody)]
+                      (:antibody.antigen/text antigen))
+          :type (if-let [antigen (:antibody/antigen antibody)]
+                    (str/capitalize (name (:antibody.antigen/value antigen))))}
+   :description "the type and decsription of antigen this antibody was generated against"})
+
+(defn constructed-by [antibody] ; no examples in datomic
```
Comment (mgrbyte) [inline src/rest_api/classes/antibody/widgets/overview.clj]: Just wondering - is the comment something that needs fixing/a TODO or just an observation?

File: src/rest_api/classes/cds/widgets/overview.clj [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+           :else
+           nil)
+   :description "Whether the start or end of the CDS is found"})
+
+(defn corresponding-all [cds] ; still trying to figure out
+  {:data (if-let [holders  (:transcript.corresponding-cds/_cds cds)]
+           (for [holder holders
+                 :let [transcript (:transcript/_corresponding-cds holder)]]
+             (let [length-unspliced (if (contains? transcript :transcript/corresponding-cds)
+                                    (let [ccds (:transcript.corresponding-cds/cds
```
Comment (mgrbyte) [inline src/rest_api/classes/cds/widgets/overview.clj]: identation looks off here (inside the if)

File: src/rest_api/classes/clone/widgets/overview.clj [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+   [clojure.string :as str]
+   [pseudoace.utils :as pace-utils]
+   [rest-api.classes.generic-fields :as generic]
+   [rest-api.formatters.date :as date]
+   [rest-api.formatters.object :as obj :refer [pack-obj]]))
+
+(defn canonical-for [clone] ; example C05B2
+  {:data (when-let [cfhs (:clone/canonical-for clone)]
+           (into {}
+             (for [cfh cfhs :let [cclone (:clone.canonical-for/clone cfh)]]
```
Comment (mgrbyte) [inline src/rest_api/classes/clone/widgets/overview.clj]: perhaps move `:let` in a `for` to it's own line for readability

File: src/rest_api/classes/construct/widgets/overview.clj [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  (:require
+   [rest-api.classes.generic-fields :as generic]
+   [rest-api.formatters.object :as obj :refer [pack-obj]]))
+
+(defn construction-summary [construct]
+  {:data (first (:construct/construction-summary construct))
+   :description "Construction details for the transgene"})
+
+(defn other-reporter [construct]
+  {:data (when-let [chs  (:construct/other-reporter construct)] (first chs))
```
Comment (mgrbyte) [inline src/rest_api/classes/construct/widgets/overview.clj]: extra space?

File: src/rest_api/classes/do_term/widgets/overview.clj [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    (for [drh drhs :let [note (:gene.disease-relevance/note drh)
+                         species (:gene.disease-relevance/species drh)]]
+      {:text note
+       :evidence (obj/get-evidence drh)})))
+
+(defn- gene-disease-orthologs [gene]
+  (when-let [ohs (:gene/ortholog gene)]
+    (remove
+      nil?
+      (for [oh ohs :let [sh (:gene.ortholog/species oh)
```
Comment (mgrbyte) [inline src/rest_api/classes/do_term/widgets/overview.clj]: `:let` on new line?

File: src/rest_api/classes/do_term/widgets/overview.clj [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+               :class "OMIM"
+               :id id})))))))
+
+(defn gene-orthology [d] ; tested on DOID:0050432 - getting more results hear than ace version
+  {:error nil
+   :data (when-let [gs (:gene.disease-potential-model/_do-term d)]
+           (vals
+             (apply
+               merge
+               (for [g gs :let [gene (:gene/_disease-potential-model g)]]
```
Comment (mgrbyte) [inline src/rest_api/classes/do_term/widgets/overview.clj]: `:let` on new line

File: src/rest_api/classes/do_term/widgets/overview.clj [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  {:data (:do-term/definition d)
+   :description "Definition of this disease"})
+
+(defn genes-biology [d] ; tested on DOID:0050432 - getting more results hear than ace version
+  {:error nil
+   :data (when-let [gs (:gene.disease-experimental-model/_do-term d)]
+           (vals
+             (apply
+               merge
+               (for [g gs :let [gene (:gene/_disease-experimental-model g)]]
```
Comment (mgrbyte) [inline src/rest_api/classes/do_term/widgets/overview.clj]: `:let` on nl

**Review by mgrbyte — COMMENTED** [review]
Generally looks ok (conflict-resolving aside)
Consider using:
```clojure
(some->> (:some-key entity)
                 (map (fn [x]
                              ....))
```
as an alternative to if-let/when-let expressions.
(just a suggestion, not a blocker)

Seems to be quite a few places where you've used `if-let` with no else condition.
I'm not adverse to either, but following you're previous statements, perhaps consider changing those that your're certain shouldn't have an else clause to use `when-let`.

File: src/rest_api/classes/anatomy_term/widgets/overview.clj [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+              :label (:anatomy-term.term/text holder)
+              :class "anatomy_name"
+              :taxonomy "all"})
+   :description "Term in the Anatomy ontology"})
+
+(defn definition [anatomy-term]
+  {:data (if-let [holder (:anatomy-term/definition anatomy-term)]
```
Comment (mgrbyte) [inline src/rest_api/classes/anatomy_term/widgets/overview.clj]: `when-let` maybe? (not sure if there's a possible else case which is why you're using if-let here)

File: src/rest_api/classes/do_term/widgets/overview.clj (hunk 1 above)
Comment (mgrbyte) [reply src/rest_api/classes/do_term/widgets/overview.clj]: @sibyl229 Usage of `some?`here is equivalent to `(if-not (nil? id)...)`

Comment (sibyl229) [reply src/rest_api/classes/do_term/widgets/overview.clj]: @mgrbyte you are right. I was making a minor point anyway. The truth value of id vs (some? id) is Only different if id is the value `false`. I don't think it's a distinction we care to make.

**Comment by sibyl229** [comment]
@a8wright In the future, could you let me know of data structure changes of the JSON during the re-write? The changes could break template. I think it's best if we could resolve template breaking changes before code freeze, as re-deploy catalyst after it goes live, while possible, isn't as easy as re-deploying d2c.

Here is an example in Analysis widget, where a scalar is replaced by an array for the description:

screen shot 2017-06-05 at 12 26 34 pm

**Review by sibyl229 — APPROVED** [review]
Ready for merge so we can test them on staging. That should give us time to uncover potential json incompatibility.

**Review by mgrbyte — COMMENTED** [review]
Generally looks ok.

~Perhaps consider using `some->>` instead of when-let/if-let + for (not a biggie, would make the code more concise perhaps)~ Sorry, already mentioned above.

**Review by mgrbyte — APPROVED** [review]

**Comment by adamjohnwright** [comment]
@sibyl229 as a formal response to your request. Yes I will try to coordinate with you when structure changes occur. As you know I, in general am trying to not make any changes from what is created by perl - that one slipped through the cracks. I think that is one of the widgets I did earlier on and had forgotten to enquire with you about the difference.

**Pull request merged**